### PR TITLE
CR-1075543 Add support for AIE dynamic profiling

### PR DESCRIPTION
**Pull request by larry9523**
This PR addresses the following issues and features
1. Fix failure of xclCloseContext that returns non-zero as successful by revert return value to 0
2. Added all other remaining 3 profiling options so that we have a full support of the 4 profiling options supported
   enum xrtProfilingOption {
      IO_TOTAL_STREAM_RUNNING_TO_IDLE_CYCLE = 0,
      IO_STREAM_START_TO_BYTES_TRANSFERRED_CYCLES,
      IO_STREAM_START_DIFFERENCE_CYCLES,
      IO_STREAM_RUNNING_EVENT_COUNT
   };
3. Re commit the "Use block mem access when read/update rtp". This was merged by https://github.com/Xilinx/XRT/pull/4093 but then reverted by https://github.com/Xilinx/XRT/pull/4096. The issue underline has been fixed and tested. We can now safely use the block mem interfaces.

@xdavidz you are welcome to review other changes, but please particularly take a look at the change in zocl_xclbin.c

### CODE REVIEW

**Review by xdavidz — APPROVED** [review]

File: src/runtime_src/core/edge/user/aie/aie.cpp [hunk 1]
```diff
@@ -392,55 +477,183 @@ start_profiling(int option, const std::string& port1_name, const std::string& po
...
 
-  Resources::AcquiredResource& acquiredResource = acquiredResourcesForThisHandle[0];
-  XAie_ModuleType XAieModuleType = AIEResourceModuletoXAieModuleTypeMap[acquiredResource.module];
+  get_profiling_config(port_name, shim_tile, mode, stream_id);
 
-  if (acquiredResource.resource == Resources::performance_counter)
-    XAie_PerfCounterGet(devInst, acquiredResource.loc, XAieModuleType, acquiredResource.id, (u32*)(&value));
-  else
-    throw xrt_core::error(-EAGAIN, "Can't read profiling: The acquired resources order does not match the profiling option.");
+  int handleId = eventRecords.size();
```
Comment (stsoe) [inline src/runtime_src/core/edge/user/aie/aie.cpp:492]: Here definitely auto should be used.  int and return type doesn't even match in this case.

File: src/runtime_src/core/edge/user/aie/aie.cpp [hunk 2]
```diff
@@ -392,55 +477,183 @@ start_profiling(int option, const std::string& port1_name, const std::string& po
...
-  return value;
+  if (counterId0 >= 0 && counterId1 >= 0 && eventPortId >= 0) {
+    XAie_EventSelectStrmPort(devInst, shim_tile, (uint8_t)eventPortId, mode, SOUTH, stream_id);
+    XAie_PerfCounterEventValueSet(devInst, shim_tile, XAIE_PL_MOD, (uint8_t)counterId1, value / 4);
+    XAie_PerfCounterControlSet(devInst, shim_tile, XAIE_PL_MOD, (uint8_t)counterId0, XAIETILE_EVENT_SHIM_PORT_RUNNING[eventPortId], XAIE_EVENT_PERF_CNT_1_PL);
+    XAie_PerfCounterControlSet(devInst, shim_tile, XAIE_PL_MOD, (uint8_t)counterId1, XAIETILE_EVENT_SHIM_PORT_RUNNING[eventPortId], XAIETILE_EVENT_SHIM_PORT_RUNNING[eventPortId]);
+    eventRecords.push_back( { IO_STREAM_START_TO_BYTES_TRANSFERRED_CYCLES,
+                { { shim_tile, Resources::pl_module, Resources::performance_counter, (size_t)counterId0 },
+                { shim_tile, Resources::pl_module, Resources::performance_counter, (size_t)counterId1 },
+                { shim_tile, Resources::pl_module, Resources::stream_switch_event_port, (size_t)eventPortId } } } );
```
Comment (stsoe) [inline src/runtime_src/core/edge/user/aie/aie.cpp:505]: Return type of various APIs and how these are used don't seem to agree very well.  Why are there so many casts?  Also C-style casting is not good.  Please use static_cast, more type yes, but more incentive to fix and easier to locate bad code.

File: src/runtime_src/core/edge/user/aie/aie.cpp [hunk 3]
```diff
@@ -392,55 +477,183 @@ start_profiling(int option, const std::string& port1_name, const std::string& po
...
+      XAie_PerfCounterControlSet(devInst, shim_tile1, XAIE_PL_MOD, (uint8_t)counterId1, XAIE_EVENT_USER_EVENT_0_PL, XAIETILE_EVENT_SHIM_PORT_RUNNING[eventPortId1]);
+      XAie_EventSelectStrmPort(devInst, shim_tile2, (uint8_t)eventPortId2, mode2, SOUTH, stream_id2);
+      XAie_PerfCounterControlSet(devInst, shim_tile2, XAIE_PL_MOD, (uint8_t)counterId2, XAIE_EVENT_USER_EVENT_0_PL, XAIETILE_EVENT_SHIM_PORT_RUNNING[eventPortId2]);
+      XAie_EventGenerate(devInst, shim_tile1, XAIE_PL_MOD, XAIE_EVENT_USER_EVENT_0_PL);
+      eventRecords.push_back({ IO_STREAM_START_DIFFERENCE_CYCLES,
+                  { { shim_tile1, Resources::pl_module, Resources::performance_counter, (size_t)counterId1 },
+                  { shim_tile2, Resources::pl_module, Resources::performance_counter, (size_t)counterId2 },
+                  { shim_tile1, Resources::pl_module, Resources::stream_switch_event_port, (size_t)eventPortId1 },
+                  { shim_tile2, Resources::pl_module, Resources::stream_switch_event_port, (size_t)eventPortId2 } } });
+      handle = handleId;
```
Comment (stsoe) [inline src/runtime_src/core/edge/user/aie/aie.cpp:555]: Is this code duplication from earlier around ~492?
Functions in this file could benefit a lot from refactoring, they are much too long and completely without comments.  Impossible to guess what's going on.

File: src/runtime_src/core/edge/user/aie/aie.cpp [hunk 4]
```diff
@@ -392,55 +477,183 @@ start_profiling(int option, const std::string& port1_name, const std::string& po
...
+  }
+
+  return handle;
 }
 
-void
+int
 Aie::
-stop_profiling(int phdl)
+start_profiling_diff_cycles(const std::string& port1_name, const std::string& port2_name)
```
Comment (stsoe) [inline src/runtime_src/core/edge/user/aie/aie.cpp:523]: Please refactor.  This function is way too long

File: src/runtime_src/core/edge/user/aie/aie.cpp [hunk 5]
```diff
@@ -392,55 +477,183 @@ start_profiling(int option, const std::string& port1_name, const std::string& po
...
+    throw xrt_core::error(-EAGAIN, "Can't start profiling: Failed to request performance counter or stream switch event port resources.");
+
+  return handle;
+}
+
+int
+Aie::
+start_profiling_event_count(const std::string& port_name)
+{
+  int handle = -1;
```
Comment (stsoe) [inline src/runtime_src/core/edge/user/aie/aie.cpp:629]: Move closer to where it is first needed.

File: src/runtime_src/core/edge/user/aie/aie.cpp [hunk 6]
```diff
@@ -392,55 +477,183 @@ start_profiling(int option, const std::string& port1_name, const std::string& po
...
+  int counterId = Resources::AIE::getShimTile(shim_tile.Col)->plModule.requestPerformanceCounter(handleId);
+
+  if (counterId >= 0 && eventPortId >= 0) {
+    XAie_EventSelectStrmPort(devInst, shim_tile, (uint8_t)eventPortId, mode, SOUTH, stream_id);
+    XAie_PerfCounterControlSet(devInst, shim_tile, XAIE_PL_MOD, (uint8_t)counterId, XAIETILE_EVENT_SHIM_PORT_RUNNING[eventPortId],                                     XAIETILE_EVENT_SHIM_PORT_RUNNING[eventPortId]);
+    eventRecords.push_back({ IO_STREAM_RUNNING_EVENT_COUNT,
+                { { shim_tile, Resources::pl_module, Resources::performance_counter, (size_t)counterId },
+                { shim_tile, Resources::pl_module, Resources::stream_switch_event_port, (size_t)eventPortId } } });
+    handle = handleId;
+  } else {
```
Comment (stsoe) [inline src/runtime_src/core/edge/user/aie/aie.cpp:648]: new line missing

**Review by stsoe — CHANGES_REQUESTED** [review]

**Comment by larry9523** [comment]
Hi @stsoe, The AIE profiling code is a temporary solution for 2020.2 so that we have full profiling support in libadf_api_xrt. In next release we will revisit the architecture, with a solid AIE resource manager (in AIE driver or somewhere), these code will be entirely refactored, probably in XDP.

**Comment by stsoe** [comment]
Merging under the assumption that this code is temporary.

**Pull request merged**